hikari/impl/shard.py

```diff
@@ -221,16 +227,18 @@ async def _receive_and_check_complete_zlib_package(
     ) -> str:
         buff = bytearray(initial_data)
 
-        while True:
+        while not buff.endswith(b"\x00\x00\xff\xff"):
             message = await self.receive(timeout)
 
```

Comment:
just a note, the old logic would've handled the CLOSE, CLOSED and CLOSING message types being received between zlib payloads but this current code treats that as a general error

Reply:
That is true, didn't even think of that, will rethink the logic.

Thanks for pointing it out

Reply:
If you want to handle those cases (that i'm not even sure could happen tbh, just wanted to point out theoretical changes in behaviour) I can just move the error code handling logic over to a function and call it in both places on error but from some cursory testing seems like payloads are rarely split between messages when using zlib transport compression

Reply:
Tbh, uppon more thinking on this, we can treat it as an error. The connection was closed but we expected a BINARY package, so error sounds good enough for me. Will test a bit and see what would actually happen if we close mid buffer.

Reply:
> that i'm not even sure could happen tbh

Cause discord, as of present, basically just doesn't send split zlib payloads so the chances of getting a close in a case that doesn't happen are like next to impossible under current behaviour but that doesn't matter since this is handling a theoretical case

Reply:
> From testing i couldn't find a case where discord splits zlib payload between payloads and have to assume its either a rare case or just doesn't ever happen. This heavily hampers the chances they'd ever be splitup by a close.

From the example in the docs, it looks like it is possible. https://github.com/discord/discord-api-docs/blob/master/docs/topics/Gateway.md#transport-compression.

> The current code would just kill the shard since GatewayError bubbles up from the shard run logic to whatever ran it

Yep, thanks for pointing it out, I didnt notice when I did it. But idk if we should change it to actually handle the close message types.

Reply:
> If it isn't reconnects wouldn't work under these cases

Ok. Will try and have a look because I have a slight hunch (random idea, probably not correct) on why https://github.com/hikari-py/hikari/blob/19a90ea879c4dcb0e61a71d7070d9f2c83df39fc/hikari/impl/shard.py#L215 is a thing. Will make sure tho, and maybe we can remove that line all together.

For now this looks good. Thanks a lot for catching it!

Reply:
I've changed the code a bit to handle closes and what not between zlib portions btw